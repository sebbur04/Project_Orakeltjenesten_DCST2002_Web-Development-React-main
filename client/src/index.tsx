import * as React from "react";
import { createRoot } from "react-dom/client";
import { useState, useEffect } from "react";
import { HashRouter, Route } from "react-router-dom";
import { createHashHistory } from "history";
import { Alert } from "./widgets";


export const history = createHashHistory(); // Use history.push(...) to programmatically change path, for instance after successfully saving a page

import userService from './service/user-service';

//Import Components for when user is logged in
import AuthTagComp from "./login/tag";
import AuthPageContent from "./login/page";
import EditPage from "./login/edit-page";
import Profile from "./login/profile";
import EditProfile from "./login/profile-edit"
import AuthSidebar from "./login/sidebar";
import NewPage from './login/new-page';
import AuthVersion from './login/version';
import AuthTopBar from './login/topbar';
import Users from './login/users'
import UserPermissions from "./login/userperm";
import Sitemap from './login/sitemap';
import AuthFooter from './login/footer';


//Import components for when user is logged out
import PageContent from './logout/page-content';
import Login from './logout/login';
import SignUp from './logout/signup';
import Sidebar from './logout/sidebar';
import TagPages from './logout/tag-pages';
import PageVersion from './logout/page-version';
import TopBar from './logout/topbar';
import sitemap from './logout/sitemap';
import Footer from './logout/footer';

//Import components that are used by both logged in and logged out
import ProfileView from './same/profile-view'
import Versionlog from './same/page-versionlog';
import AllPages from './same/all-pages';
import Main from './same/main';


//Help from ChatGPT to write code with states here
export function useAuth() {
//ChatGPT har hjulpet til med states her
  const [user, setUser] = useState({ id: 0, username: "", avatar: "", bio: "", permid: 0}); // Initialize with default values

  useEffect(() => {
    async function fetchUser() {
      try {
        const response = await userService.getUserAuth(); // Call the API to get the current user
        if (response && response.user) {
          // Update state with the fetched user data
          setUser({ id: response.user.id, username: response.user.username, avatar: response.user.avatar, bio: response.user.bio, permid: response.user.permid });
        }
      } catch (err) {
        console.log(err); // Handle error (e.g., user not authenticated)
      }
    }
    fetchUser(); // Fetch user data on component mount
  }, []); // Run once on component mount

  return { user }; // Return user
}

//Returns different HashRouters if you're authenticated or not. Has the added bonus of stopping a logged out user from accessing any of the components used by a logged in user. 
function App(){

  const authUser = useAuth();

  if(authUser.user.id != 0){ { /* LOGGED IN */}
    //Only return authenticated components / components for authenticated users.
    return(
      <HashRouter>
        <Route path="*" component={AuthTopBar} />
        <div className="page-container">
          <AuthSidebar />
          <div className="mainContent">

          <Alert/>
          <Route exact path="/" component={Main} />
          <Route exact path="/tags/:id" component={AuthTagComp} />
          <Route exact path="/pages/:id" component={AuthPageContent} />
          <Route exact path="/pages/:id/new" component={NewPage} />
          <Route exact path="/pages/:id/edit" component={EditPage} />
          <Route exact path="/pages/:id/versionlog" component={Versionlog} />
          <Route exact path="/pages/:pageid/versionlog/:versionnum" component={AuthVersion} />
          <Route exact path="/profile" component={Profile} /> 
          <Route exact path="/useredit" component={EditProfile} /> 
          <Route exact path="/users/:id" component={ProfileView} />
          <Route exact path="/allpages" component={AllPages} /> 
          <Route exact path="/users" component={Users} />
          <Route exact path="/users/:id/permissions" component={UserPermissions} />
          <Route exact path="/sitemap" component={Sitemap} />
        </div></div>
        <Route exact path="*" component={AuthFooter} />
      </HashRouter>
    )
  } else {
    //Return only components for users that are not authenticated. 
    return(
      <HashRouter>
        <Route path="*" component={TopBar} />
        <div className="page-container">
          <Sidebar />
          <div className="mainContent">
          <Alert/>
          <Route exact path="/" component={Main} />
          <Route exact path="/tags/:id" component={TagPages} />
          <Route exact path="/pages/:id" component={PageContent} />
          <Route exact path="/login" component={Login} />
          <Route exact path="/signup" component={SignUp}/>
          <Route exact path="/pages/:id/versionlog" component={Versionlog} />
          <Route exact path="/pages/:pageid/versionlog/:versionnum" component={PageVersion} />
          <Route exact path="/allpages" component={AllPages} /> 
          <Route exact path="/users/:id" component={ProfileView} />
          <Route exact path="/sitemap" component={sitemap} />
        </div>

        </div>
        <Route exact path="*" component={Footer} />
      </HashRouter>
    );
  }
}

export default App;

// Rendering the Application
let root = document.getElementById("root");
if (root)
  createRoot(root).render(<App />);
