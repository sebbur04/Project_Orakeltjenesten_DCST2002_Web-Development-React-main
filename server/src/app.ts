// Import express and express-session
import express from "express";
import session from "express-session";

// Import routers
import tagRouter from "./router/tag-router";
import userRouter from "./router/user-router";
import pageRouter from "./router/page-router";
import authRouter from "./router/auth-router";
import commentRouter from "./router/comment-router";
import permissionRouter from "./router/permission-router";


var passport = require("passport");
var path = require("path");

//Used to store session data. Since this should not be sent in plain text to our mysql database.
var SQLiteStore = require("connect-sqlite3")(session);
/**
 * Express application.
 */
const app = express();

app.use(express.json());

app.use(express.static(path.join(__dirname, "public")));

app.use(
  session({
    secret: "keyboard cat",
    resave: false,
    saveUninitialized: false,
    store: new SQLiteStore({ db: "sessions.db", dir: "./var/db" }),
  })
);

app.use(passport.initialize());
app.use(passport.session());
app.use(passport.authenticate("session"));

// Since API is not compatible with v1, API version is increased to v2
// Added routes for all routers based on our setup

app.use("/api/v2/tags", tagRouter);
app.use("/api/v2/users", userRouter);
app.use("/api/v2/pages", pageRouter);
app.use("/api/v2/comments", commentRouter);
app.use("/api/v2/permissions", permissionRouter);
app.use("/api/v2/", authRouter);

export default app;
