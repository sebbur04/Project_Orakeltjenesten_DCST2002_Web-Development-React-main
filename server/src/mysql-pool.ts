// Import mysql and sqlite3
import mysql from "mysql2";
import sqlite3 from "sqlite3";
const mkdirp = require("mkdirp");

// To set up a folder with sqlite3 database. Retrieved from Passportjs tutorial for local login.
mkdirp.sync("./var/db");

// SQLite: used only for session storage
const sqliteDB = new sqlite3.Database("./var/db/sessions.db");

/**
 * MySQL connection pool with options specified in the following environment variables:
 * - `MYSQL_HOST`
 * - `MYSQL_USER`
 * - `MYSQL_PASSWORD`
 * - `MYSQL_DATABASE`
 */
const pool = mysql.createPool({
  host: process.env.MYSQL_HOST,
  user: process.env.MYSQL_USER,
  password: process.env.MYSQL_PASSWORD,
  database: process.env.MYSQL_DATABASE,
  // Reduce load on NTNU MySQL server
  connectionLimit: 1,
  // Convert MySQL boolean values to JavaScript boolean values
  typeCast: (field, next) =>
    field.type == "TINY" && field.length == 1 ? field.string() == "1" : next(),
});

export { pool, sqliteDB };
