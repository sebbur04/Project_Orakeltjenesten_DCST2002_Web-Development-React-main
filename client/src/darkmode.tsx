import * as React from "react";
import { createRoot } from "react-dom/client";
import { Component } from "react-simplified";
import { Card, Form, Row, Button, Alert } from "./widgets"; 
import { useState, useEffect } from "react";
import { NavLink, HashRouter, Route, useParams } from "react-router-dom";
import { createHashHistory } from "history"; // Brukes for å lagre darkmode state lokalt, slik at den ikke endres hver gang man går til en ny side, eller
                                             // refresher siden.                         

// Dark Mode Button Component
export default function DarkModeButton() {
  // Lagrer darkMode state i localStorage. URL:https://www.freecodecamp.org/news/how-to-use-localstorage-with-react-hooks-to-set-and-get-items/
  // Bruker localStorage.getItem('darkMode') for å hente darkMode state fra localStorage. Hvis det ikke finnes en state, settes den til false.
  const [darkMode, setDarkMode] = useState(() => {
    const savedMode = localStorage.getItem('darkMode');
    return savedMode === 'true'; // sjekker om darkMode er lagret fra den forrige sesjonen. Hvis den er det, returneres true, ellers false.
  });

  const toggleDarkMode = () => {
    const newMode = !darkMode;
    setDarkMode(newMode); // Oppdaterer darkMode state
    document.body.classList.toggle('dark-mode', newMode); // Bytter mellom dark-mode og light-mode klassen på body
    localStorage.setItem('darkMode', newMode.toString()); // Lagrer darkMode state i localStorage
  };


  useEffect(() => {
    if (darkMode) {
      document.body.classList.add('dark-mode');
    }
  }, [darkMode]); // Dependency array. Kjører useEffect hver gang darkMode endres. URL: https://react.dev/reference/react/useEffect#specifying-reactive-dependencies

  // Render av dark mode button
  return (
    <button className="dark-mode-button" onClick={toggleDarkMode}>
      <img
        src={darkMode ? '/light-mode-icon.png' : '/dark-mode-icon.png'} // Lightmode icon hvis darkmode er true, og omvendt
        alt={darkMode ? 'Light Mode' : 'Dark Mode'} // Hvis iconet ikke dukker opp, vises teksten "Light Mode" eller "Dark Mode". Ble brukt i starten for å se om iconet ble lastet inn. 
        style={{ width: '30px', height: '30px' }} // Størrelse på ikonet
      />
    </button>
  );
}