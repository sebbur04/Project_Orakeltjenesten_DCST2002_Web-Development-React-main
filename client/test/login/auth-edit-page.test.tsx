// Import neccessary modules for testing
import * as React from "react";
import { act } from "react";
import { shallow } from "enzyme";
import Edit from "../../src/login/edit-page";
import { Button, Form, Column, Row } from "../../src/widgets";

// Mocking services with the mock structure
jest.mock("../../src/service/page-service", () => ({
  getPage: jest.fn(() =>
    Promise.resolve({ id: 1, name: "Test Page", num_views: 10 })
  ),
  getTagsFromPage: jest.fn(() => Promise.resolve([{ id: 1, name: "Tag1" }])),
  updatePageName: jest.fn(() => Promise.resolve()),
  deletePage: jest.fn(() => Promise.resolve()),
  getAllPages: jest.fn(() => Promise.resolve([{ id: 1, name: "Test Page" }])),
}));

jest.mock("../../src/service/versionlog-service", () => ({
  getLatestVersion: jest.fn(() =>
    Promise.resolve({
      id: 1,
      pageid: 1,
      content: "Sample content",
      version: 1,
      changelog: "Initial version",
    })
  ),
  createVersion: jest.fn(() => Promise.resolve()),
}));

jest.mock("../../src/service/user-service", () => ({
  getAuthenticatedUser: jest.fn(() =>
    Promise.resolve({
      id: 1,
      username: "User1",
      avatar: "avatar",
      bio: "bio",
      permid: 1,
    })
  ),
  getAllUsers: jest.fn(() =>
    Promise.resolve([
      { id: 1, username: "User1", avatar: "avatar", bio: "bio", permid: 1 },
    ])
  ),
}));

jest.mock("../../src/service/tag-service", () => ({
  getAllTags: jest.fn(() =>
    Promise.resolve([
      { id: 1, name: "Tag1" },
      { id: 2, name: "Tag2" },
    ])
  ), // Return a tag
}));

jest.mock("../../src/service/permission-service", () => ({
  getPermission: jest.fn(() =>
    Promise.resolve({
      id: 1,
      alterpages: true,
      deletepages: true,
      versions: true,
      allcomments: true,
      tags: true,
      users: true,
    })
  ),
}));

jest.mock("../../src/service/page-tags-service", () => ({
  addTagToPage: jest.fn(() => Promise.resolve()),
  deleteTagFromPage: jest.fn(() => Promise.resolve()),
}));

describe("Edit Component Tests", () => {
  test("renders correctly and displays the page title and form fields (snapshot)", async () => {
    const wrapper = shallow(<Edit match={{ params: { id: 1 } }} />);

    // Wait for asynchronous operations to complete
    //@ts-ignore
    await wrapper.instance().mounted();
    wrapper.update();

    // Assert that components are rendered correctly
    expect(wrapper).toMatchSnapshot();
  });

  test("Displays correctly", async () => {
    const wrapper = shallow(<Edit match={{ params: { id: 1 } }} />);

    // Wait for asynchronous operations to complete
    //@ts-ignore
    await wrapper.instance().mounted();
    wrapper.update();

    expect(
      wrapper.containsAllMatchingElements([
        <Row>
          <Column width={1}>Tittel </Column>
          <Column width={2}>
            <Form.Input
              width="20vw"
              maxlength={25}
              type="text"
              value="Test Page"
              onChange={() => {}}
            />
          </Column>
        </Row>,
        <Row>
          <Column width={2}>
            <Button.Success onClick={() => {}}>Legg til tagg</Button.Success>
          </Column>
        </Row>,
        <Row>
          <Column width={2}>Endringslogg:</Column>
          <Column width={6}>
            <Form.Input
              width="50vw"
              type="text"
              value="Initial version"
              onChange={() => {}}
            />
          </Column>
        </Row>,
        <Row>
          <Column>Generer lenke: </Column>
          <Column>
            <Form.Checkbox checked={false} onChange={() => {}} />
          </Column>
        </Row>,
      ])
    );
  });

  test("handles adding and removing tags correctly", async () => {
    let wrapper;

    // Wrap async operations in act
    await act(async () => {
      wrapper = shallow(<Edit match={{ params: { id: 1 } }} />); // Use mount for full rendering

      // Ensure that the component is fully rendered
      expect(wrapper.exists()).toBe(true); // Ensure the wrapper is not empty

      // Wait for async operations (e.g., fetching permission and tags)
      //@ts-ignore
      await wrapper.instance().mounted(); // Ensure async methods like mounted() are finished
      wrapper.update(); // Ensure the component re-renders after async updates
    });

    // Check if tags are loaded properly
    //@ts-ignore
    const addButton = wrapper.find(Button.Success).at(0);
    expect(addButton.exists()).toBe(true); // Ensure the add button exists

    // Simulate adding a tag (button click)
    await act(async () => {
      addButton.simulate("click");
    });

    // Wait for the component to update after adding the tag
    //@ts-ignore
    await wrapper.update();

    // Assert that the Form.Select (or whatever element is used for tag selection) is now rendered
    //@ts-ignore
    expect(wrapper.find(Form.Select).length).toBe(1);

    // Check for the remove button and simulate clicking it
    //@ts-ignore
    const removeButton = wrapper.find(Button.Danger);
    expect(removeButton.exists()).toBe(true); // Ensure the remove button exists

    // Simulate the remove button click
    await act(async () => {
      removeButton.at(0).simulate("click");
    });

    // Wait for the component to update after removing the tag
    //@ts-ignore
    await wrapper.update();

    // Assert that the Form.Select (or similar element) is no longer present
    //@ts-ignore
    expect(wrapper.find(Form.Select).length).toBe(1);
  });

  test("handles the save button click (with valid inputs)", async () => {
    // Mock response for tags if it's not already mocked globally
    jest.mock("../../src/service/tag-service", () => ({
      getAllTags: jest.fn(() => Promise.resolve([{ id: 1, name: "Tag1" }])), // Mocking tags service
    }));

    const wrapper = shallow(<Edit match={{ params: { id: 1 } }} />);

    // Wait for asynchronous operations to complete (like data fetching)
    //@ts-ignore
    await wrapper.instance().mounted(); // Assuming `mounted()` is where data fetching occurs
    wrapper.update(); // Ensure component is updated after the async operation

    // Set values for the form fields
    wrapper
      .find(Form.Input)
      .at(0)
      .simulate("change", { currentTarget: { value: "Updated Page Name" } });
    wrapper
      .find(Form.Input)
      .at(1)
      .simulate("change", { currentTarget: { value: "Updated Changelog" } });

    // Simulate first button click (save)
    wrapper.find(Button.Success).at(1).simulate("click"); // Simulate second button click (save)

    setTimeout(() => {
      // Expect the location to change
      expect(location.hash).toEqual("#/");
    });
  });

  test("handles the delete button click", async () => {
    const wrapper = shallow(<Edit match={{ params: { id: 1 } }} />);

    // Wait for asynchronous operations to complete
    //@ts-ignore
    await wrapper.instance().mounted();
    wrapper.update();

    // Simulate deleting the page
    wrapper.find(Button.Danger).at(1).simulate("click");

    // Wait for async operation and hash change
    await wrapper.update();
    expect(location.hash).toEqual("#/");
  });
});
