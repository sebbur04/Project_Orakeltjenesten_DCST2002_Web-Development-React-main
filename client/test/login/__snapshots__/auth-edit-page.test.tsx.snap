// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit Component Tests renders correctly and displays the page title and form fields (snapshot) 1`] = `
<div
  className="mainContent"
>
  <Card
    title="Rediger side"
  >
    <Row>
      <Column
        width={1}
      >
        Tittel 
      </Column>
      <Column
        width={2}
      >
        <FormInput
          maxlength={25}
          onChange={[Function]}
          type="text"
          value="Test Page"
          width="50vw"
        />
      </Column>
    </Row>
    <Row>
      <Column>
        <b>
          Tagger
        </b>
      </Column>
    </Row>
    <Row
      key="1"
    >
      <Column
        width={3}
      >
        Tag1
      </Column>
      <Column
        width={2}
      >
        <ButtonDanger
          onClick={[Function]}
        >
          Fjern tagg
        </ButtonDanger>
      </Column>
    </Row>
    <Row>
      <Column
        width={3}
      >
        <FormSelect
          onChange={[Function]}
          value={2}
        >
          <option
            key="2"
            value={2}
          >
            Tag2
          </option>
        </FormSelect>
      </Column>
      <Column
        width={2}
      >
        <ButtonSuccess
          onClick={[Function]}
        >
          Legg til tagg
        </ButtonSuccess>
      </Column>
    </Row>
    <Row>
      <Column
        width={2}
      >
        Endringslogg:
      </Column>
      <Column
        width={6}
      >
        <FormInput
          onChange={[Function]}
          placeholder="Skriv hva du har endret her"
          type="text"
          value=""
          width="50vw"
        />
      </Column>
    </Row>
    <Row>
      <Row>
        <Column
          width={2}
        >
          Generer lenke: 
        </Column>
        <Column>
          <FormCheckbox
            checked={false}
            onChange={[Function]}
          />
        </Column>
      </Row>
      <Row>
        <div
          className="wide-textarea"
        >
          <textarea
            className="textarea"
            onChange={[Function]}
            rows={10}
            value="Sample content"
          />
        </div>
      </Row>
    </Row>
    <Row>
      <Column
        width={7}
      >
        <ButtonDanger
          onClick={[Function]}
        >
          Slett denne siden
        </ButtonDanger>
      </Column>
      <Column
        width={2}
      >
        <ButtonLight
          onClick={[Function]}
        >
          Avbryt
        </ButtonLight>
      </Column>
      <Column
        width={2}
      >
        <ButtonSuccess
          onClick={[Function]}
        >
          Lagre
        </ButtonSuccess>
      </Column>
    </Row>
  </Card>
  <Card
    title="ForhÃ¥ndsvisning"
  >
    <Row>
      <Qe>
        ---
      </Qe>
    </Row>
    <Row>
      <Qe>
        Sample content
      </Qe>
    </Row>
  </Card>
</div>
`;
